### PR TITLE
Add support to `GradScaler` for respecting an already set `grad_scale` value 

### LINKED ISSUE #123428: Change `GradScaler` to respect an existing `grad_scale` value.
### 🚀 The feature, motivation and pitch

**Background**
The `grad_scale` value supported by various fused optimizers (e.g.[ Adam](https://github.com/pytorch/pytorch/blob/cfd06bd60cc959622292112ba535a2aadc58f7c7/torch/optim/adam.py#L187) and AdamW) allows a fused kernel to divide the gradients by a denominator before computing the update. The[ `GradScaler` uses this to unscale the gradients](https://github.com/pytorch/pytorch/blob/cfd06bd60cc959622292112ba535a2aadc58f7c7/torch/amp/grad_scaler.py#L435-L437) before computing the optimizer updates in a single kernel. 

**Pitch**
There are conceivable reasons why a user, outside of the `GradScaler` might want to do the same thing. For example, to support efficient custom gradient scaling.

**Existing behaviour**
The issue is the `GradScaler` will override the value of `grad_scale` that a user may have already set. 

**Proposal**
My proposal is simple, if `grad_scale` is already set to `x` and the `GradScaler` has a scale of `y`, we simply set `grad_scale=x*y` before the optimizer step.

In code, this would change 
https://github.com/pytorch/pytorch/blob/cfd06bd60cc959622292112ba535a2aadc58f7c7/torch/amp/grad_scaler.py#L435-L437 
to 
```python
optimizer.grad_scale = (  # type: ignore[attr-defined]
    None if optimizer_state["stage"] == OptState.UNSCALED else scaler * getattr(optimizer, "grad_scale", 1) 
)
``` 
This is safe since post `optimizer.step()` the value gets deleted and so it won't cause the grad_scale to be multiplied again and again.

If this sounds good to you, I've raised a PR [here](https://github.com/pytorch/pytorch/pull/123429)!

### Alternatives

_No response_

### Additional context

_No response_

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar @mcarilli @ptrblck @leslie-fang-intel @jgong5

### PR DESCRIPTION
**Pull request by yousufmo**
Fixes #123428

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
needs a test

**Comment by ezyang** [comment]
i'll land this as is but can you sign CLA?

**Comment by yousufmo** [comment]
Apologies for the delay - I've signed the CLA and I'm happy to add tests. Would `test/test_cuda.py` be the right place?

I've also corrected the implementation to address the following 
```python
import torch
import torch.nn as nn
from torch.cuda.amp import GradScaler
from torch.optim import Adam

device = torch.device(0)
W = nn.Parameter(torch.ones((5,5), device=device))
x = torch.ones(5, device=device)
yhat = torch.zeros(5, device=device)
optimizer = Adam(params=[W], fused=True)
scaler = GradScaler()
y = W@x
l = ((y-yhat)**2).mean()
l = scaler.scale(l)
l.backward()

scaler.unscale_(optimizer)

# unscaled gradients are all two
assert torch.allclose(W.grad, torch.full_like(W, fill_value=2))

# the user sets a `grad_scale` value which should be fused with the optimizer step
optimizer.grad_scale = torch.Tensor([10]).cuda()
scaler.step(optimizer)

# we now expect the gradient to have been in-place modified by the Adam step and divided by 10
assert torch.allclose(W.grad, torch.full_like(W, fill_value=0.2))
```

Without `getattr(optimizer, "grad_scale", None)` calling `unscale_` would result in ignoring the user-supplied `grad_scale`.

**Comment by ezyang** [comment]
I don't see a test_amp.py, so I suppose cuda is as good a place as any.

**Comment by yousufmo** [comment]
@ezyang - this is now ready for re-review. I added the tests and checked they pass. 

Sorry about the multiple committers - they are in fact both me! Have signed the CLA from both accounts.
